pkg/compactor/block_upload.go

```diff
@@ -0,0 +1,399 @@
+	"github.com/grafana/mimir/pkg/util"
+	util_log "github.com/grafana/mimir/pkg/util/log"
+)
+
+const tmpMetaFilename = "uploading-" + block.MetaFilename
+
+var rePath = regexp.MustCompile(`^(index|chunks/\d{6})$`)
+
+// HandleBlockUpload handles requests for starting or completing block uploads.
+//
+// The query parameter uploadComplete (true or false, default false) controls whether the
+// upload should be completed or not.
+//
+// Starting the uploading of a block means to upload meta file and verify that the upload can go ahead.
+// In practice this means to check that the (complete) block isn't already in block storage,
+// and that meta file is valid.
+func (c *MultitenantCompactor) HandleBlockUpload(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	blockID := vars["block"]
+	if blockID == "" {
+		http.Error(w, "missing block ID", http.StatusBadRequest)
+		return
+	}
+	bULID, err := ulid.Parse(blockID)
+	if err != nil {
+		http.Error(w, "invalid block ID", http.StatusBadRequest)
+		return
+	}
+	tenantID, ctx, err := tenant.ExtractTenantIDFromHTTPRequest(r)
+	if err != nil {
+		http.Error(w, "invalid tenant ID", http.StatusBadRequest)
+		return
+	}
+
+	logger := util_log.WithContext(ctx, c.logger)
+	logger = log.With(logger, "block", blockID)
+
+	userBkt := bucket.NewUserBucketClient(tenantID, c.bucketClient, c.cfgProvider)
+	if err := checkForComplete(ctx, logger, bULID, userBkt); err != nil {
+		var httpErr httpError
+		if errors.As(err, &httpErr) {
+			level.Warn(logger).Log("msg", httpErr.message, "err", err)
+			http.Error(w, httpErr.message, httpErr.statusCode)
+			return
+		}
+
+		level.Error(logger).Log("msg", "an unexpected error occurred", "err", err)
+		http.Error(w, "internal server error", http.StatusInternalServerError)
+		return
+	}
+
+	if r.URL.Query().Get("uploadComplete") == "true" {
+		err = c.completeBlockUpload(ctx, r, logger, userBkt, tenantID, bULID)
+	} else {
+		err = c.createBlockUpload(ctx, r, logger, userBkt, tenantID, bULID)
+	}
+	if err != nil {
+		var httpErr httpError
+		if errors.As(err, &httpErr) {
+			level.Warn(logger).Log("msg", httpErr.message, "err", err)
+			http.Error(w, httpErr.message, httpErr.statusCode)
+			return
+		}
+
+		level.Error(logger).Log("msg", "an unexpected error occurred", "err", err)
+		http.Error(w, "internal server error", http.StatusInternalServerError)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+}
+
+// checkForComplete checks for a complete block with same ID. If one exists, an error is returned.
+func checkForComplete(ctx context.Context, logger log.Logger, blockID ulid.ULID, userBkt objstore.Bucket) error {
+	exists, err := userBkt.Exists(ctx, path.Join(blockID.String(), block.MetaFilename))
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("failed to check existence of %s in object storage", block.MetaFilename))
+	}
+	if exists {
+		level.Debug(logger).Log("msg", "complete block already exists in object storage")
```

Comment:
No need to debug-log this. Caller will already warn-log the message from httpError, which is the same.